app/com/arpnetworking/metrics/portal/scheduling/JobExecutorActor.java

```diff
@@ -18,171 +18,486 @@
+        final JobRef<T> typedRef = ref;
+        return typedRef;
+    }
+
+    private void attemptExecuteAndUpdateRepository(final Instant scheduled) {
+        if (!_cachedJob.isPresent()) {
+            killSelf();
+            throw new IllegalStateException("unable to execute: executor is not initialized");
+        }
+
+        final CachedJob<T> cachedJob = _cachedJob.get();
+
+        final JobRef<T> ref = cachedJob.getRef();
+        final Job<T> job = cachedJob.getJob();
+        final JobRepository<T> repo = ref.getRepository(_injector);
+
+        if (_currentlyExecuting.getAndSet(true)) {
+            return;
+        }
+
+        try {
+            repo.jobStarted(ref.getJobId(), ref.getOrganization(), scheduled);
+        } catch (final NoSuchElementException error) {
+            _currentlyExecuting.set(false);
+            killSelf();
+            LOGGER.warn()
+                    .setMessage("attempted to start executing job, but job no longer exists in repository")
+                    .addData("ref", cachedJob.getRef())
+                    .addData("scheduled", scheduled)
+                    .log();
+            return;
+        }
+
+        PatternsCS.pipe(
+                job.execute(getSelf(), scheduled)
+                        .handle((result, error) -> new JobCompleted.Builder<T>()
+                                    .setScheduled(scheduled)
+                                    .setOutcome(error, result)
+                                    .build()),
+                getContext().system().dispatcher()
+        ).to(getSelf());
+    }
+
+    private void tick(final Tick message) {
+        _periodicMetrics.recordCounter("job_executor_actor_ticks", 1);
+
+        if (!_cachedJob.isPresent()) {
+            killSelf();
+            throw new IllegalStateException("somehow, uninitialized JobExecutorActor is trying to tick");
+        }
+        final CachedJob<T> cachedJob = _cachedJob.get();
+
+        final Optional<Instant> nextRun = cachedJob.getSchedule().nextRun(cachedJob.getLastRun());
+        if (!nextRun.isPresent()) {
+            LOGGER.info()
+                    .setMessage("job has no more scheduled runs")
+                    .addData("cachedJob", cachedJob)
+                    .log();
+            killSelf();
+            return;
+        }
+
+        if (_clock.instant().isBefore(nextRun.get().minus(EXECUTION_SLOP))) {
+            scheduleTickFor(nextRun.get());
         } else {
-            return TICK_INTERVAL;
+            attemptExecuteAndUpdateRepository(nextRun.get());
         }
     }
 
-    private void scheduleNextTick(final Instant wakeUpBy) {
-        timers().startSingleTimer("TICK", Tick.INSTANCE, timeUntilNextTick(wakeUpBy));
+    private void reload(final Reload<T> message) {
+        final Optional<String> eTag = message.getETag();
+        final CachedJob<T> cachedJob;
+        _periodicMetrics.recordCounter("job_executor_actor_reloads", 1);
```

Comment:
This seems redundant because the cached actor is already instrumented for the same. See comment above on it. One avenue could be to put the instrumentation only here. Not sure if that helps.

Reply:
Maybe I went a bit metrics-happy. This certainly is, information-theoretically, redundant, because any aggregate of this metric can be computed from aggregates of the reload / conditional-reload metrics. But I don't like writing MetaQueries in Grafana. Could still remove this if you feel strongly.

Reply:
No, that's fine. Just remember changing metric names, definitions or removing them is effectively changing the "operational interface" of the service. Graphs, alerts and runbooks can depend on these. Metrics should have clear intention, naming and therefore obvious meaning to operators (the hallmarks of a good "interface"). Not suggesting this doesn't meet that bar, just figured I'd point and ask. Leaving it is fine.